src/app/CommandResponseSender.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "CommandResponseSender.h"
+#include "InteractionModelEngine.h"
+#include "messaging/ExchangeContext.h"
+
+namespace chip {
+namespace app {
+using Status = Protocols::InteractionModel::Status;
+
+CHIP_ERROR CommandResponseSender::OnMessageReceived(Messaging::ExchangeContext * apExchangeContext,
+                                                    const PayloadHeader & aPayloadHeader, System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err                     = CHIP_NO_ERROR;
+    bool sendStatusResponseWithFailure = false;
+
+    if (mState == State::AwaitingStatusResponse &&
+        aPayloadHeader.HasMessageType(Protocols::InteractionModel::MsgType::StatusResponse))
+    {
+        CHIP_ERROR statusError = CHIP_NO_ERROR;
+        err                    = StatusResponse::ProcessStatusResponse(std::move(aPayload), statusError);
+        VerifyOrExit(err == CHIP_NO_ERROR, sendStatusResponseWithFailure = true);
+        err = statusError;
+        VerifyOrExit(err == CHIP_NO_ERROR, sendStatusResponseWithFailure = true);
+
+        err = SendCommandResponse();
```

Comment:
So here's the key thing that needs to be documented.  If this _fails_, the exchange will auto-close because we are processing an incoming message on it and don't send a response.  That's what makes it safe to not close exchanges in SendCommandResponse: the only callers are here and StartSendingCommandResponses, and the latter expects its caller to handle closing the exchange....

All of this is very subtle and needs to be very clearly documented.

Reply:
Hopefully latest comments is enough

Reply:
I would have said:
```
        // If SendCommandResponse() fails, we are responsible for closing the exchange,
        // as stipulated by the API contract. We fulfill this obligation by not sending
        // a message expecting a response on the exchange.  Since we are in the middle
        // of processing an incoming message, the exchange will close itself once we are
        // done processing it, if there is no response to wait for at that point.
```